Makes special role preferences a list, instead of flag entry.

PsiOmegaDelta opened the pull request:
The list of available roles in the character setup is now automatically generated based on available antag and special roles, instead of being hardcoded.
Updates some role usages to use the new ghost trap system.

:cl:
tweak: Antagonist and special role preferences has been overhauled. Please update these specific character preferences as they have been reset.
/:cl:

PsiOmegaDelta committed:
Makes special role preferences a list, instead of flag entry.

Allows us to define more than 16 special roles.
Updates some role usages to use the new ghost trap system.

GinjaNinja32 commented on code/modules/client/preference_setup/antagonism/02_candidacy.dm:
```diff
@@ -1,54 +1,51 @@
-		var/num = text2num(href_list["be_special"])
-		pref.be_special ^= (1<<num)
+	if(href_list["add_special"])
+		pref.be_special_role |= href_list["add_special"]
```
should likely verify the validity of the selected option here

GinjaNinja32 commented on code/modules/ghosttrap/trap.dm:
```diff
@@ -58,9 +61,18 @@ proc/populate_ghost_traps()
 	if(href_list["candidate"] && href_list["target"])
 		var/mob/dead/observer/candidate = locate(href_list["candidate"]) // BYOND magic.
 		var/mob/target = locate(href_list["target"])                     // So much BYOND magic.
+		var/valid_until = text2num(href_list["valid_until"])
```
Trusting the href?

Manual input of hrefs is entirely possible, don't trust them.

PsiOmegaDelta replied:
I am aware, just didn't consider it at all important for this case, but very well.

GinjaNinja32 commented on code/modules/ghosttrap/trap.dm:
```diff
@@ -111,4 +128,47 @@ proc/populate_ghost_traps()
+* Cordical Borer *
+*****************/
+/datum/ghosttrap/borer
+	object = "cordical borer"
```
cortical

GinjaNinja32 commented on code/modules/mob/living/carbon/brain/posibrain.dm:
```diff
@@ -16,63 +16,14 @@
 	if(brainmob && !brainmob.key && searching == 0)
 		//Start the process of searching for a new user.
-		user << "\blue You carefully locate the manual activation switch and start the positronic brain's boot process."
+		user << "<span class='notice'>You carefully locate the manual activation switch and start the positronic brain's boot process.</b>"
```
`<span class='notice'>` ... `</b>`???

GinjaNinja32 commented on code/modules/mob/living/simple_animal/borer/borer.dm:
```diff
@@ -173,41 +173,8 @@
-	into the head of its victim. Use stealth, persuasion and your powers of mind control to keep you, \
-	your host and your eventual spawn safe and warm."
-	src << "You can speak to your victim with <b>say</b>, to other borers with <b>say :x</b>, and use your Abilities tab to access powers."
+	var/datum/ghosttrap/G = get_ghost_trap("cordical borer")
```
cortical again

PsiOmegaDelta committed:
Adds the necessary infrastructure to not have to trust the href request timeout for ghost trap requests.

Skips unregistering.

GinjaNinja32 commented on code/modules/ghosttrap/trap.dm:
```diff
@@ -46,10 +55,19 @@ proc/populate_ghost_traps()
 		if(O.client)
-			O << "[request_string]<a href='?src=\ref[src];candidate=\ref[O];target=\ref[target]'>Click here</a> if you wish to play as this option."
+			if(request_timeout)
+				request_timeouts[target] = world.time
```
`world.time + request_timeout`?

PsiOmegaDelta committed:
Tweaks.

Re-uses the too_many_active_drones() proc as originally intended.
Makes the private_valid_special_roles list global, no need to let each individual player initialize the list.

GinjaNinja32 commented on code/modules/ghosttrap/trap.dm:
```diff
@@ -46,10 +55,19 @@ proc/populate_ghost_traps()
-			O << "[request_string]<a href='?src=\ref[src];candidate=\ref[O];target=\ref[target]'>Click here</a> if you wish to play as this option."
+			if(request_timeout)
+				request_timeouts[target] = world.time + request_timeout
+				target.destruction.register(src, /datum/ghosttrap/proc/target_destroyed)
```
This entire block should likely be called outside the `for(var/mob/dead/observer/O in player_list)`

GinjaNinja32 commented on code/modules/mob/living/living.dm:
```diff
@@ -814,8 +814,11 @@ default behaviour is:
-/mob/living/proc/can_be_possessed_by(var/mob/dead/observer/possessor)
-	if(!istype(possessor))
+/mob/proc/can_be_possessed_by(var/mob/dead/observer/possessor)
+	return istype(possessor) && possessor.client  && possessor.ckey
```
technically don't need the `ckey` check here, if `client` is non-null, then `ckey` is guaranteed to be non-null.

GinjaNinja32 commented:
:+1: with Travis.

PsiOmegaDelta commented:
Given that this is untested, as per the original post, I wouldn't recommend merging it.

GinjaNinja32 commented:
:+1: with travis and testing then :P

PsiOmegaDelta committed:
Makes all special role entries use unique ids now that it is possible.

PsiOmegaDelta commented:
Additional changes implemented and basic testing conducted.
